File: Graph/all_topological_sorts.cpp
#include<iostream>
#include<vector>

using namespace std;


void print_graph(int** graph, int v);
void all_topological_sorts(int **graph, int v, bool* visited, vector<int>& result, int *indegree);


/*
Initialize all vertices as unvisited.
Now choose vertex which is unvisited and has zero indegree and decrease indegree of all those vertices by 1 (corresponding to removing edges) now add this vertex to result and call the recursive function again and backtrack.
After returning from function reset values of visited, result and indegree for enumeration of other possibilities.
*/

int main() {
	int n=0;
	cout << "Enter the number of vertices: ";
	cin >> n;
	int** graph;
	graph=new int*[n];
	for(int i=0;i<n;i++) {
		graph[i]=new int[n];
	}
	//Graph Initialization
	cout << "Enter the edges: " <<endl;
	int x;
	while(true) {
		cout << "\n\n******* MENU *******" <<endl;
		cout <<"Press \'1\' to enter edge." <<endl;
		cout <<"Press \'2\' to print the current graph." <<endl;
		cout <<"Press \'3\' to exit entering edge." <<endl;
		cin >> x;
		if(x==3) {
			break;
		}
		else if(x==2) {
			print_graph(graph,n);
		}
		else if(x==1) {
			int u,v;
			cout << "Enter the first vertex of the edge: ";
			cin >> u;
			cout << "Enter the second vertex of the edge: ";
			cin >>v;
			if(u<0||v<0||u>=n||v>=n) {
				cout << "Invalid Input. Try Again!" << endl;
				continue;
			}
			graph[u][v]=1;
		}
		else {
			cout << "Invalid Input. Try Again!" <<endl;
			continue;
		}
	}
	print_graph(graph,n);
	// Visited Boolean Array
	bool* visited=new bool[n];
	// Indegree Array
	int* indegree=new int[n];
	//Calculating indegree
	for(int i=0;i<n;i++) {
		indegree[i]=0;
	}
	for(int i=0;i<n;i++) {
		for(int j=0;j<n;j++) {
			if(graph[i][j]==1) {
				indegree[j]+=1;
			}
		}
	}
	vector<int> result;
	cout << "\n\nAll Topological Sorts: " <<endl;
	all_topological_sorts(graph,n,visited,result, indegree);
}


void print_graph(int** graph, int v) {
	cout <<"\n\nCurrent Graph: " <<endl;
	for(int i=0;i<v;i++) {
		for(int j=0;j<v;j++) {
			cout << graph[i][j] << "\t";
		}
		cout << endl;
	}

}

void all_topological_sorts(int **graph, int v, bool* visited, vector<int>& result, int *indegree) {
	bool flag=false;
	//For each vertex of the graph
	for(int i=0;i<v;i++) {
		//If the vertex is not visited and has 0 indegree.
		if(!visited[i]&&indegree[i]==0) {
			//Pushed to result.
			result.push_back(i);
			//Lowering the indegree of neighbouring nodes of current node by 1.
			for(int j=0;j<v;j++) {
				if(graph[i][j]==1) {
					indegree[j]-=1;
				}
			}
			//Marking current node as visited.
			visited[i]=true;
			//Recursive call
			all_topological_sorts(graph,v,visited,result,indegree);
			//Removed from result.
			result.erase(result.end() - 1); 
			//Marking current node as unvisited.
			visited[i]=false;
			//Increasing the indegree of neighbouring nodes of current node by 1.
			for(int j=0;j<v;j++) {
				if(graph[i][j]==1) {
					indegree[j]+=1;
				}
			}
			//Marking the flag true to print the order.
			flag=true;
		}
	}
	if (!flag) { 
        for (int i = 0; i < result.size(); i++) {
            cout << result[i] << " "; 
        }
        cout << endl; 
    }	
}

File: src/java_codes/graph/AllTopologicalSorts.java
package java_codes.graph;

import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;

/*
Initialize all vertices as unvisited.
Now choose vertex which is unvisited and has zero indegree and decrease indegree of all those vertices by 1 (corresponding to removing edges) now add this vertex to result and call the recursive function again and backtrack.
After returning from function reset values of visited, result and indegree for enumeration of other possibilities.
*/

public class AllTopologicalSorts {

  private final int v;
  private final List<List<Integer>> adjList;

  public AllTopologicalSorts(int v) {
    this.v = v;
    adjList = new ArrayList<>(v);
    for (int i = 0; i < v; i++) {
      adjList.add(new ArrayList<>());
    }
  }

  public void addEdge(int u, int v) {
    adjList.get(u).add(v);
  }

  public void allTopologicalSorts() {
    boolean[] visited = new boolean[v];
    int[] indegree = new int[v];
    for (int i = 0; i < v; i++) {
      for (int to : adjList.get(i)) {
        indegree[to]++;
      }
    }

    List<Integer> result = new ArrayList<>();
    System.out.println("\nAll Topological Sorts: ");
    allTopologicalSortsUtil(visited, result, indegree);
  }

  private void allTopologicalSortsUtil(boolean[] visited, List<Integer> result, int[] indegree) {
    boolean flag = false;
    for (int i = 0; i < v; i++) {
      if (!visited[i] && indegree[i] == 0) {
        visited[i] = true;
        result.add(i);
        for (int to : adjList.get(i)) {
          indegree[to]--;
        }

        allTopologicalSortsUtil(visited, result, indegree);

        visited[i] = false;
        result.remove(result.size() - 1);
        for (int to : adjList.get(i)) {
          indegree[to]++;
        }
        flag = true;
      }
    }
    if (!flag) {
      for (int i : result) {
        System.out.print(i + " ");
      }
      System.out.println();
    }
  }

  public static void main(String[] args) {
    Scanner scanner = new Scanner(System.in);
    System.out.print("Enter the number of vertices: ");
    int n = scanner.nextInt();
    AllTopologicalSorts g = new AllTopologicalSorts(n);

    while (true) {
      System.out.println("\n******* MENU *******");
      System.out.println("Press '1' to enter edge.");
      System.out.println("Press '2' to find all topological sorts.");
      System.out.println("Press '3' to exit.");
      int x = scanner.nextInt();
      if (x == 3) {
        break;
      } else if (x == 2) {
        g.allTopologicalSorts();
      } else if (x == 1) {
        System.out.print("Enter the first vertex of the edge: ");
        int u = scanner.nextInt();
        System.out.print("Enter the second vertex of the edge: ");
        int v = scanner.nextInt();
        if (u < 0 || v < 0 || u >= n || v >= n) {
          System.out.println("Invalid Input. Try Again!");
          continue;
        }
        g.addEdge(u, v);
      } else {
        System.out.println("Invalid Input. Try Again!");
      }
    }
    scanner.close();
  }
}
